---
title: "seurat_cc_deg_calc"
output: html_document
date: "2022-08-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "~/Coding/R/scrnaseq")
library(Seurat)
library(SeuratData)

# library(devtools)
# devtools::install_github('mojaveazure/seurat-disk')
library(SeuratDisk)

# devtools::install_github("theislab/zellkonverter")
library(zellkonverter)

library(dplyr)
library(ggplot2)
library(utils)
```

# Calculating Cell Cycle/Differential Gene Expression with Seurat

This is an R markdown notebook created to calculate cell cycle and differential gene expression scores through Seurat. It takes a .h5ad (ANNDATA) object as input.

## STEP 0: Prepare Env and Get Data
#### set global parameters
```{r}
# CHANGE THESE AS NECESSARY
adata_in <- "D:/Data/setd1a/5-seurat/in/10_01_2023_setd1a_b4.log1p_liger_med_singleR_filter.h5ad"
out_dir <- "D:/Data/setd1a/5-seurat/out"

prefix <- "10_01_2023_b4_" # i.e. (date)_(dataset)_
cc_out <- paste(prefix, "cc_out", sep = "")
deg_out <- paste(prefix, "deg_out", sep = "")

# parameters for Cell Cycle Scoring
cc_score_thresh <- 50
desired_cc_cell_types <- c("RG", "IPC")

# parameters for Differential Gene Expression Scoring
deg_thresh <- 100
```

#### convert desired .h5ad into a Seurat object using `zellkonverter` package
```{r}
# read in the desired dataset as an h5ad to prepare for zellkonverter
setd1a_sce <- readH5AD(adata_in, verbose = TRUE)

# convert the h5ad into a Seurat object using zellkonverter pkg
setd1a <- as.Seurat(setd1a_sce, counts = "X", data = NULL)
```

#### initialise and preprocess the new Seurat object
This step is done in accordance with the vignettes on the Seurat site.
``` {r}
setd1a <- NormalizeData(setd1a)
setd1a <- FindVariableFeatures(setd1a, selection.method = "vst")
setd1a <- ScaleData(setd1a, features = rownames(setd1a))

setd1a <- RunPCA(setd1a, features = VariableFeatures(setd1a), ndims.print = 6:10, nfeatures.print = 10)
```

## STEP 1: Calculate Cell Cycle Score

#### determine the desired cell types to calculate scores for
There are two parameters that can be changed to conduct a slightly different analysis:
* `cc_score_thresh`: determines how many occurrences a cell type needs to be included in analysis (default 50)
* `desired_cc_cell_types`: determines which types of cells (e.g. RG, IPC, EN, IN, etc.) should be included in analysis
``` {r}
cell_types <- c()

# initialise a dataframe with values set as the names of the cells sampled
cell_type_all <- as.data.frame(setd1a$nowakowski.noglyc)

# ensure that only cell types with count above cc_score_thresh are kept
cell_type_counts <- cell_type_all %>% count(setd1a$nowakowski.noglyc)
cell_type_counts <- cell_type_counts[cell_type_counts$n > cc_score_thresh, ]
cell_type_distinct <- cell_type_counts %>% select("setd1a$nowakowski.noglyc")

# select out the desired cell types and assign them into a vector for use later
for (i in cell_type_distinct$`setd1a$nowakowski.noglyc`) {
  desired = FALSE
  
  for (j in desired_cc_cell_types) {
    if (grepl(j, i, fixed = TRUE) == TRUE) {
      desired = TRUE
    }
  }
  if (desired == TRUE) {
    cell_types <- c(cell_types, i)
  }
}
```

#### initialise an empty master cc score dataframe and create empty out dir
``` {r}
master_df <- data.frame()

dir.create(file.path(paste(out_dir, cc_out, sep="/")), showWarnings = FALSE)
```


#### calculate cc scores
``` {r}
for (i in cell_types) {
  # load Seurat's included cc markers
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  
  # subset data by cell type
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc == i)
  
  # run cell cycle score analysis
  setd1a_sub <- CellCycleScoring(setd1a_sub, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  
  # save plots
  pdf(file = paste(paste(out_dir, cc_out, sep="/"), i, "_cc_score_RidgePlot.pdf", sep = ""))
  print(
    RidgePlot(setd1a_sub, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)
  )
  dev.off()
  
  pdf(file = paste(paste(out_dir, cc_out, sep="/"), "/", i, "_cc_score_DimPlot.pdf", sep = ""))
  print(
    DimPlot(setd1a_sub)
  )
  dev.off()

  # save data to csv and add it to the master score df
  df <- as.data.frame(setd1a_sub[[]])
  master_df <- master_df %>% bind_rows(df)
 
  write.csv(df, paste(paste(out_dir, cc_out, sep="/"), "/", i, "_cc_score_data.csv", sep = ""), row.names = TRUE)
}
```

#### save master cc score df to a csv
``` {r}
write.csv(master_df, paste(paste(out_dir, cc_out, sep="/"),"/all_cc_score_data.csv", sep = ""), row.names = TRUE)
```

STEP 2: calculate differential gene scores

#### determine the desired cell types to calculate scores for
There is one parameter that can be changed to conduct a slightly different analysis:
* `deg_thresh`: determines how many occurrences a cell type needs to be included in analysis (default 100)
``` {r}
cell_types <- c()

cell_type_all <- as.data.frame(setd1a$nowakowski.noglyc)

cell_type_counts <- cell_type_all %>% count(setd1a$nowakowski.noglyc)
cell_type_counts <- cell_type_counts[cell_type_counts$n > deg_thresh, ]
cell_type_distinct <- cell_type_counts %>% select("setd1a$nowakowski.noglyc")
```

#### initialise empty out dir
``` {r} 
dir.create(file.path(paste(out_dir, deg_out, sep="/")), showWarnings = FALSE)
```

#### calculate diff gene exp scores with MAST

NOTE: When calculating diff gene exp scores, it is necessary to set the identities of the object to the `sampleID` column in order for the comparisons between different samples to be made. The identities can be changed as desired by using `Idents(object = [your Seurat object]) <- [desired data column]`.

``` {r}
for (i in cell_type_distinct$`setd1a$nowakowski.noglyc`) {
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc == i)
  
  # set identities to compare between samples
  Idents(object = setd1a_sub) <- setd1a_sub$sampleID
  Idents(object = setd1a_sub)

  # install.packages("BiocManager")
  
  # run these if first time on local pc
  # BiocManager::install("limma")
  # BiocManager::install("MAST")
  # BiocManager::install("DESeq2")
  
  # save MAST score
   fs3_cntrl3_mast <- FindMarkers(setd1a_sub, ident.1 = "FS3", ident.2 = "CT3",
                                 test.use = "MAST", logfc.threshold = 0, only.pos = FALSE)
   write.csv(fs3_cntrl3_mast, paste(paste(out_dir, deg_out, sep="/"), "/", i, "_fs3_cntrl3_mast.csv", sep = ""), row.names = TRUE)
} 
```


#### calculate diff gene exp scores with DESeq2
WARNING: DESeq2 requires integral datapoints, and may throw an error if there are non-integers present.
``` {r}
for (i in cell_type_distinct$`setd1a$nowakowski.noglyc.noglyc`) {
  
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc.noglyc == i)
  setd1a_sub[[]]
  
  # set identities to compare between samples
  Idents(object = setd1a_sub) <- setd1a_sub$sampleID
  Idents(object = setd1a_sub)
  
  # install.packages("BiocManager")
  
  # run these if first time on local pc
  # BiocManager::install("limma")
  # BiocManager::install("MAST")
  # BiocManager::install("DESeq2")
  
  # save DESeq2 score
  fs3_cntrl3_deseq2 <- FindMarkers(setd1a_sub, ident.1 = "FS3", ident.2 = "CT3", test.use = "DESeq2")
  write.csv(fs3_cntrl3_deseq2, paste(paste(out_dir, deg_out, sep="/"), "/", i, "_fs3_cntrl3_deseq2.csv", sep = ""), row.names = TRUE)
} 
```